Minimal support of floating-point operand bundles

spavloff opened the pull request:
This changes modifies the semantics of floating-point intrinsics that represent math functions. Rather than using different intrinsics for default and strictfp modes, the same intrinsic is used in both cases. In strictfp function, such an intrinsic call acquires the side effects necessary to guarantee proper instruction ordering. Additionally, floating-point operations may be annotated with operand bundles, which provide supplemental information to the compiler. With the operand bundles, the new semantics offer a full-fledged representation of floating-point operations suitable for capturing their behavior in strictfp functions.

The purpose is to obtain a representation of floating-point operations, flexible enough to cover features that cannot currently be expressed by using constrained functions, like static rounding, denormal behavior and so on. The motivations and ideas behind this work are described in the discussion https://discourse.llvm.org/t/rfc-change-of-strict-fp-operation-representation-in-ir/85021.

The modified semantics are designed to work in parallel with the existing constrained functions. Both mechanisms use the same underlying model based on memory effects and should not interfere with each other's operation. Additionally, the floating-point operations represented as instructions, not calls, cannot have operand bundles, they must be converted to alias call first. So constrained functions are currently used to represent such operations.

spavloff committed:
Minimal support of floating-point operand bundles

This is a lite version of https://github.com/llvm/llvm-project/pull/109798,
where code changes are minimized to facilitate discussion about the
implementation. The motivations and ideas behind the new floating-point
operation support are described in that PR and in the discussion
https://discourse.llvm.org/t/rfc-change-of-strict-fp-operation-representation-in-ir/85021.
There are concerns that the proposed changes are too invasive and a new
approach is required to make the transition smoother.

This implementation is essentially a subset of PR109798, where
everything beyond the minimum is removed. It tries to build eventually
the same implementation as that PR but in different steps.

The patch does not attempt to modify the existing implementation based
on the constrained intrinsics. Instead it introduces a new one using
operand bundles. This new implementation initially has very limited
functionality, which latter will be extended and finally can replace the
existing one.

This PR introduces the notion of floating-point operation, this is an
intrinsic, that is listed in the file "FloatingPointOps.def". These have
two additional properties:

1. In the strict environment (a function with strictfp attribute) calls
   to these operations acquire side effect, now it is read/write access
   to inaccessible memory, just as constrained intrinsics do.

2. Calls to these operations may have floating-point operand bundles.
   There are two kinds of such bundles, tagged with "fp.control" and
   "fp.except", which are used to carry additional information about
   control modes and exception handling. Initially the set of control
   modes consists of rounding mode only.

The set of operations enlisted in "FloatingPointOps.def" and in
"ConstrainedOps.def" are completely independent, an intrinsic may be in
one list or in both. The set of floating-point operations is expected to
grow and finally all FP intrinsics will be available in the new
implementation. In this patch set of intrinsics in
"FloatingPointOps.def" is minimum necessary for tests.

spavloff committed:
Set MemoryEffects of call only if called function is nomem

It make call close to constrained calls, where memory effects of the
calls do not define memory effects. It helps migration to operand
bundles.

spavloff committed:
Fix method signature

spavloff committed:
Organize checks in Verifier

dtcxzyw reviewed (commented):
I don't see the necessity of using operand bundles. All of the possible arguments are metadata strings, not SSA values.
Can't we just use metadata instead?
```
call float @llvm.nearbyint.f32(float %x), !fp.control !{ !"dyn"}, !fp.except !{ !"strict"}
```

kpneal commented:
> I don't see the necessity of using operand bundles. All of the possible arguments are metadata strings, not SSA values. Can't we just use metadata instead?

Operand bundles are replacing the constrained intrinsics so we can use the bundles on the hundreds (thousands?) of intrinsics that deal with floating point. Because of this, we can't create new intrinsics that take additional metadata strings.

There's also a desire to add more floating point state to the bundles later. For example, denormal flushing on/off.  We can't do that with metadata strings now, and I haven't heard of any method to add them later. Operand bundles seems to be our best bet.

nikic commented:
> I don't see the necessity of using operand bundles. All of the possible arguments are metadata strings, not SSA values. Can't we just use metadata instead?

Metadata always has to be droppable. Some of this could in theory be handled with metadata if we say the behavior (with strictfp) is strict by default and the metadata relaxes it. But this doesn't work for everything. For example it would not work for specifying a static rounding mode, as dropping that metadata changes the semantics of the operation.

spavloff committed:
Small changes in documentation

jcranmer-intel reviewed (commented):
Tests for FP operand bundles on a non-intrinsic function would be helpful.

spavloff commented:
> Tests for FP operand bundles on a non-intrinsic function would be helpful.

This patch's operand bundles are designed specifically for use with intrinsic functions, similar to constrained intrinsics. Using operand bundles with user-supplied function would require refining semantic of this use.

spavloff commented:
Thre is a separate fix for lowering in Selection DAG: https://github.com/llvm/llvm-project/pull/138553,

spavloff committed:
Update according to the obtained feedback

- Bundles are renamed: "fp.control" -> "fp.round",
- Corrected documentation,
- DAG nodes STRICT_* are used to lower the supported intrinsics,
- Constrained properties in IRBuilder are reused for handling FP bundles,
- DAG gets methods for lowering FP operations,
- IR dump prints FP memory effects,
- Fixed some problems in IRBuilderBase::CreateCall,
- Added a new check to Verifier,
- Added tests for checking code generation.

spavloff committed:
Address review comments

- Add IRBuilder::resetModeToStrictFP instead of adding extra argument to
  setIsFPConstrained.
- Change wording in the Release Notes.
- Modify printing MemoryEffects of FP operation.
- Fix IRBuilder::CreateCall.
- Clean up unit test.

spavloff commented on clang/lib/CodeGen/CodeGenFunction.cpp:
```diff
@@ -1092,7 +1092,7 @@ void CodeGenFunction::StartFunction(GlobalDecl GD, QualType RetTy,
       (!FD && (FPExceptionBehavior != llvm::fp::ebIgnore ||
                RM != llvm::RoundingMode::NearestTiesToEven))) {
-    Builder.setIsFPConstrained(true);
+    Builder.setIsFPConstrained(true, false);
```
In addition to `setIsFPConstrained`, which keeps original signature, a method `resetModeToStrictFP(bool IsCon)` is added, which also resets IRBuilder's FP properties.

spavloff commented on llvm/docs/ReleaseNotes.md:
```diff
@@ -56,6 +56,8 @@ Makes programs 10x faster by doing Special New Thing.
 Changes to the LLVM IR
 ----------------------
 
+* Floating-point operand bundles have been added.
```
Provided a bit more explanation.

spavloff commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -9355,6 +9357,56 @@ bool SelectionDAGBuilder::visitBinaryFloatCall(const CallInst &I,
+bool SelectionDAGBuilder::visitFPOperation(const CallInst &I, unsigned Opcode) {
+  // We already checked this call's prototype; verify it doesn't modify errno.
+  MemoryEffects ME = I.getMemoryEffects();
+  if (!ME.onlyAccessesInaccessibleMem())
```
It wouldn't make the check more correct. Actually the test for `onlyAccessesInaccessibleMem` is also excessive, as the call in this method is a call to an extern function, intrinsics are handles in another place. External functions have memory effects as "reads and writes all memories".

spavloff committed:
Fix warnings

spavloff committed:
Put back 'maytrap' and fix some errors

spavloff committed:
Allow more intrinsics to have FP bundles

With this change all intrinsics that have constrained counterparts are
allowed to have floating-point operand bundles as well.

spavloff committed:
Fix copying math flags

spavloff committed:
When ilniling do not convert calls with bundles

spavloff committed:
Do not set fp.round for functions that do not depend on rounding mode

spavloff committed:
Support of assumed rounding mode

spavloff committed:
Inlining instructions with FP bundles

spavloff committed:
Cleanup

spavloff committed:
Fix behavior in Alias Analysis

spavloff committed:
Fix tests after rebase

spavloff committed:
Fix setting NoFPExcept flag

spavloff committed:
Restore test

spavloff committed:
Fix format errors

spavloff committed:
Updated documentation

spavloff committed:
Fix typo

spavloff committed:
Remove unnecessary changes
